poetry/core/factory.py

```diff
@@ -134,6 +159,154 @@ def create_poetry(self, cwd=None):  # type: (Optional[Path]) -> Poetry
+                    file_path,
+                    category=category,
+                    base=root_dir,
+                    extras=constraint.get("extras", []),
+                )
+            elif "path" in constraint:
+                path = Path(constraint["path"])
+
+                if root_dir:
+                    is_file = root_dir.joinpath(path).is_file()
+                else:
+                    is_file = path.is_file()
+
+                if is_file:
+                    dependency = FileDependency(
+                        name,
+                        path,
+                        category=category,
+                        optional=optional,
+                        base=root_dir,
+                        extras=constraint.get("extras", []),
+                    )
+                else:
+                    dependency = DirectoryDependency(
+                        name,
+                        path,
+                        category=category,
+                        optional=optional,
+                        base=root_dir,
+                        develop=constraint.get("develop", True),
+                        extras=constraint.get("extras", []),
+                    )
+            elif "url" in constraint:
+                dependency = URLDependency(
+                    name,
+                    constraint["url"],
+                    category=category,
+                    optional=optional,
+                    extras=constraint.get("extras", []),
+                )
+            else:
+                version = constraint["version"]
+
+                source_info = {}
+                if "source" in constraint:
+                    source_info = self.get_source_information(constraint["source"])
+
+                dependency = Dependency(
+                    name,
+                    version,
+                    optional=optional,
+                    category=category,
+                    allows_prereleases=allows_prereleases,
+                    extras=constraint.get("extras", []),
+                    **source_info
+                )
+
+            if not markers:
+                marker = AnyMarker()
+                if python_versions:
+                    dependency.python_versions = python_versions
+                    marker = marker.intersect(
+                        parse_marker(
+                            create_nested_marker(
+                                "python_version", dependency.python_constraint
+                            )
+                        )
+                    )
+
+                if platform:
+                    marker = marker.intersect(
+                        parse_marker(
+                            create_nested_marker(
+                                "sys_platform", parse_generic_constraint(platform)
+                            )
+                        )
+                    )
+            else:
+                marker = parse_marker(markers)
```

Comment:
`markers` will always have priority over `python` and `platform` so they are indeed mutually exclusive.

Regarding the propagation of markers, I prefer we keep the current implementation, for now at least, since it's the most accurate way to have platform agnostic lock files.

We are not ready to have self sufficient lock files and it would most likely need to rethink the current design.

Reply:
I am still not fully convinced about that rationale as this is not necessarily about being platform agnostic, but rather making sure any markers specified for root dependencies are propagated down. I do not believe this will chnage the status quo per say. Ofcourse I might be missing some pieces here. 

That said I am happy to take this particular discussion elsewhere.

As for being mutually exclusing I have a feeling this does have UX consequences. Since this is the existing behaviour we should simply doc it in another PR to avoid the confusion.